scripts/restore-typescript-cache.js

```diff
@@ -0,0 +1,87 @@
+const {execSync} = require('child_process');
+const fs = require('fs-extra');
+
+const {exec, gracefulExit} = require('./utilities');
+
+process.on('uncaughtException', gracefulExit);
+process.on('unhandledRejection', gracefulExit);
+
+const resetModifiedTimesCommand = 'touch -m -t 201001010000';
+
+/*
+ * This script is an ugly hack to speed up CI time :/  It is necessary because:
+ * - TypeScript's incremental build caching is based on file modification dates
+ * - Git always clones repos using the current date/time for all files
+ * - So in CI, TypeScript will always view a cache as 100% "stale"
+ *
+ * To work around this, the script:
+ * - Resets the modification dates of all config / src files to a time in the
+ *   far past
+ * - Grabs all changed paths for the branch
+ * - Tweaks the modification date of all changed paths
+ *
+ * This gets the cache / git in sync, albeit in a very gross manner 🤭
+ */
+
+async function run() {
+  const branchName = process.env.TRAVIS_PULL_REQUEST_BRANCH;
+  if (!branchName || branchName === 'master') {
+    console.log('Skipping build cache for master');
```

Comment:
Outside the scope of this PR, but got me thinking a bit. 

Locally, running the cold cache builds are painfully long. If CI uploads master build caches to a _thing_ and locally `yarn run build` pulls the cache from _thing_, would that work to bring down cold-cache local builds.

Reply:
Yeah, that's possible.